### PR TITLE
login button

### PR DESCRIPTION
**Pull request by AnthonyClarke25**
test

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,6 +9,8 @@
     "lint": "next lint"
   },
   "dependencies": {
+    "@clerk/backend": "^0.37.0",
```
Comment (DMCFA) [inline package.json]: why did you add these 2?

File: src/app/page.tsx [hunk 2]
```diff
@@ -1,6 +1,10 @@
+import Navbar from "@/components/Navbar";
+
 export default function Home() {
   return (
-    <main className='flex min-h-screen flex-col items-center justify-between p-24'>
+    <main className="flex min-h-screen flex-col items-center justify-between p-24">
+      <Navbar />
```
Comment (DMCFA) [inline src/app/page.tsx:6]: this should go into the root layout if it's always visible (even for signed out users/guests) or on the dashboard one if it's only for users

File: src/components/button.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import Link from "next/link";
+import React, { Children } from "react";
+
+interface Props {
```
Comment (DMCFA) [inline src/components/button.tsx:4]: I would also add a type property with limited options (button, submit, reset etc)

File: src/components/button.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  url?: string;
+  clickFn?: () => void;
+  children: React.ReactNode;
+  btnclass: string;
+}
+
+const Button = ({ url, clickFn, children, btnclass }: Props) => {
```
Comment (DMCFA) [inline src/components/button.tsx]: We will probably need to add forward Ref here to keep the element ref in re renders - we can look at it later

File: src/components/button.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import Link from "next/link";
```
Comment (DMCFA) [inline src/components/button.tsx]: we should add 'use client' here for the on click?

File: src/components/Navbar.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import React from "react";
```
Comment (DMCFA) [inline src/components/Navbar.tsx]: Is this import needed? Next should already do this behind the scenes

**Review by DMCFA — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (AnthonyClarke25) [reply package.json]: i think i added this back when i was trying to get the enviorment to run and the suggestion i got from searching online was to make sure i had the latest clerk installed. I can remove

File: src/app/page.tsx (hunk 2 above)
Comment (AnthonyClarke25) [reply src/app/page.tsx:6]: i added this here just so i could see it as we haven't defined anything yet

File: src/components/button.tsx (hunk 5 above)
Comment (AnthonyClarke25) [reply src/components/button.tsx]: might need an explaination for this one

Comment (AnthonyClarke25) [reply src/components/button.tsx]: nevermind did a google

File: package.json (hunk 1 above)
Comment (DMCFA) [reply package.json]: Can you remove and try to run? The clerk next package should already have this but I might be wrong so please test it

File: src/components/Navbar.tsx [hunk 7]
```diff
@@ -1,5 +1,6 @@
-import React from "react";
-import Button from "./Button";
+"usse client";
```
Comment (DMCFA) [inline src/components/Navbar.tsx]: typo but let's remove it from here and try to use it locally only. This will improve performance as nav will still be a server component

File: src/components/button.tsx [hunk 8]
```diff
@@ -1,20 +1,28 @@
...
   clickFn?: () => void;
   children: React.ReactNode;
   btnclass: string;
+  type?: "button" | "submit" | "reset";
 }
 
-const Button = ({ url, clickFn, children, btnclass }: Props) => {
+const Button = ({ url, clickFn, children, btnclass, type }: Props) => {
+  let buttonRef = useRef<HTMLButtonElement>(null);
+  let linkRef = useRef<HTMLAnchorElement>(null);
```
Comment (DMCFA) [inline src/components/button.tsx]: since you only need one ref at a time (either for the button or the link), you can use a single ref and dynamically assign it based on the condition

File: src/components/button.tsx [hunk 9]
```diff
@@ -1,20 +1,28 @@
...
+
   const component = url ? (
-    <Link href={url} className={btnclass}>
-      {children}
+    <Link href={url} passHref legacyBehavior>
+      <a ref={linkRef} className={btnclass}>
```
Comment (DMCFA) [inline src/components/button.tsx]: reg previous comment
`<a ref={linkRef as RefObject<HTMLAnchorElement>}`

**Review by DMCFA — CHANGES_REQUESTED** [review]

File: src/components/button.tsx [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+const Button = ({ url, clickFn, children, btnclass, type }: Props) => {
+  let buttonRef = useRef<HTMLButtonElement>(null);
+  let linkRef = useRef<HTMLAnchorElement>(null);
+
+  const component = url ? (
+    <Link href={url} passHref legacyBehavior>
+      <a ref={linkRef} className={btnclass}>
+        {children}
+      </a>
+    </Link>
```
Comment (DMCFA) [inline src/components/button.tsx]: just regarding the link comment, I was trying to understand why you are using legacy behavior but the docs are not very clear so I wasn't sure of what's correct either, but I have tried to create something like the below and it worked for me so can you just clarify so I can learn what's happening here?

see file attached for mock button component
[button.txt](https://github.com/bookshift/bookshift/files/13854887/button.txt)

Anyway I'll accept your solution, just want to understand why.

Comment (AnthonyClarke25) [reply src/components/button.tsx]: you can't pass ref to a Link. so the only way suggested is to have an a tag within the link. However, that is deprecated,as of nextjs 10 I think , so you need to use legacy to enable it.

File: src/components/button.tsx (hunk 8 above)
Comment (AnthonyClarke25) [reply src/components/button.tsx]: i intially tried to do ref = useref<htmlbutton| htmlanchort>(null) but the complied wouldnt acceptit. using chat gpt gave me this as a suggestion

File: src/components/Navbar.tsx (hunk 7 above)
Comment (AnthonyClarke25) [reply src/components/Navbar.tsx]: i will need to check it again but in nextjs if you don't specify it's for a client the default behaviour is to assume it's for a server so you are able to have client children on a use server. or at least it wouldn't let me when I tried to run it from just the nav component.

Comment (DMCFA) [reply src/components/Navbar.tsx]: I think you should be able to remove as long as the button component is a client component

File: src/components/button.tsx (hunk 8 above)
Comment (DMCFA) [reply src/components/button.tsx]: that's correct so if the compile didn't accept it your issue was probably on my next comment line. 
got to be careful with gpt as sometimes it will give you a lazy or false answer

File: src/components/button.tsx (hunk 9 above)
Comment (DMCFA) [reply src/components/button.tsx]: refer to the comment above

File: src/components/button.tsx (hunk 10 above)
Comment (DMCFA) [reply src/components/button.tsx]: do you have a source for this as I couldn't find in their comments and I tested the file attached where I pass the ref directly. I will log the ref later today to see if I'm missing something

File: package.json [hunk 11]
```diff
@@ -16,7 +16,8 @@
     "@prisma/client": "^5.7.1",
     "next": "14.0.4",
     "react": "^18",
-    "react-dom": "^18"
+    "react-dom": "^18",
+    "react-icons": "^4.12.0"
```
Comment (DMCFA) [inline package.json]: fyi you'll probably run into a conflict here as I'm already using on another branch off main

File: src/components/button.tsx [hunk 12]
```diff
@@ -1,33 +1,25 @@
...
-
-import Link from "next/link";
 import React, { useRef } from "react";
 
 interface Props {
-  url?: string;
   clickFn?: () => void;
   children: React.ReactNode;
-  btnclass: string;
   type?: "button" | "submit" | "reset";
```
Comment (DMCFA) [inline src/components/button.tsx]: this shouldn't be optional, a button should always have a type

**Review by DMCFA — CHANGES_REQUESTED** [review]
just change the type and you can merge it in
